config option to override method argument and return type nullability

carlossless opened the pull request:
This PR partially resolves https://github.com/madsmtm/objc2/issues/753#issuecomment-2871949083 by adding options in translation-config.toml to override the nullability of method argument and return types (including properties)

madsmtm commented on crates/header-translator/src/config.rs:
```diff
@@ -582,6 +606,9 @@ pub struct MethodData {
     pub unsafe_: Unsafe,
+    #[serde(rename = "override")]
+    #[serde(default)]
+    pub override_: Option<MethodOverride>,
```
Nit: I'd avoid this struct altogether, everything in here is an "override". Just inline `pub return_: Option<ReturnOverride>,` here (or maybe `pub return_: ReturnOverride`, since it impls Default).

madsmtm commented on crates/header-translator/src/config.rs:
```diff
@@ -572,6 +572,30 @@ pub struct CategoryData {
+    #[default]
+    Nullable,
+    NonNull,
+    Unspecified,
```
Nit: Wouldn't add this as a "user-facing" option

madsmtm commented on crates/header-translator/src/method.rs:
```diff
@@ -535,28 +535,29 @@ impl Method {
+            documentation: Documentation::from_entity(&entity, context),
+        };
+
+        method.apply_config_overrides(&data);
```
I'd like this to happen _before_ construction of `Method`, and I'd like it to also happen when constructing properties. Maybe make `apply_config_overrides` a free-standing function instead?

madsmtm reviewed (commented):
Short review, am at RustWeek so a bit limited time. But generally looks fine

carlossless commented:
@madsmtm I addressed your points and extended this override to properties and method arguments. Please let me know what you think and please give me pointers on how this should be tested, thanks!

madsmtm commented on crates/header-translator/src/method.rs:
```diff
@@ -909,3 +931,19 @@ pub(crate) fn handle_reserved(name: &str) -> String {
+        };
+
+        match ty {
+            Ty::Pointer { nullability, .. } => {
+                *nullability = c_nullability;
+            }
```
`Ty::Sel` and `Ty::IncompleteArray` has null-ability too. Could I get you to handle that as well?

Also, it'd be nice to emit an `warning!` for "nullabilities are equal / redundantly set in `translation-config.toml`".

carlossless replied:
Yep, addressed :+1:

madsmtm commented on framework-crates/objc2-io-usb-host/translation-config.toml:
```diff
@@ -5,9 +5,9 @@ custom-lib-rs = true
-class.IOUSBHostControllerInterface.methods."getPortStateMachineForPort:error:".skipped = true
+# Return type nullability is incorrectly declared in the header. It should be `nullable` instead of `nonnull`.
+class.IOUSBHostControllerInterface.methods."getPortStateMachineForCommand:error:".return.nullability = "nullable"
+class.IOUSBHostControllerInterface.methods."getPortStateMachineForPort:error:".return.nullability = "nullable"
```
Nit: Could I get you to re-run the translator, and commit the result? You don't have to open a PR to `objc2-generated`, as long as you've pushed it to GH then I can find it when merging.

If it's too bothersome, then I can do it for you?

carlossless replied:
Sure, no problem, it's here https://github.com/carlossless/objc2-generated/commit/b8d328c65b921049a9faacf6b97256aad501417f

madsmtm replied:
I meant, you have to link it in the commit in this PR. I think something like `git add generated`

carlossless replied:
Oh, I wasn't aware this could work this way. Is this what you mean?

madsmtm replied:
Yup, exactly.

madsmtm reviewed (approved):
Thanks, looks fine now, only minor comments

carlossless commented:
@madsmtm addressed your last comments. Also, let me know if you prefer to change the warning message to something else!

madsmtm commented:
Thanks! Will merge once CI finishes
